### ISSUE TITLE
Improper usage of sys.exit()

### ISSUE THREAD

**Issue by catdog2** [sosy-lab/benchexec#102]
I noticed that, especially in runexecutor.RunExecutor, sys.exit() is used to handle error cases quite often. It's fine to simply exit the interpreter from code which is only meant to be run standalone but inside code which may be imported this seems to be very improper. Third party code should be given a chance to handle such cases properly by raising something derived from Exception (you could catch SystemExit but this feels just wrong here).

**Comment by PhilippWendler** [comment]
I agree. I do not know though which kind of exception should typically be used to signal problems coming from invalid user input or an environment that is not suitable for the given task. What exception would you suggest?

**Comment by catdog2** [comment]
For invalid function arguments [ValueError](https://docs.python.org/2.7/library/exceptions.html#exceptions.ValueError) or [TypeError](https://docs.python.org/2.7/library/exceptions.html#exceptions.TypeError) seem to be the obvious choice.
 
[EnvironmentError](https://docs.python.org/2/library/exceptions.html#exceptions.EnvironmentError) looks like a good choice if the environment is not suitable.

**Comment by PhilippWendler** [comment]
Both `ValueError` and `TypeError` would typically indicate a bug, though, wouldn't they?

It seems like catching them would be much worse than catching `SystemExit` because it would not distinguish between bugs in RunExecutor and invalid arguments, and thus hide bugs?

Which cases do you actually want to catch? What is your use case for this? I think that if the caller does everything correctly, RunExecutor should never raise an exception.

**Comment by catdog2** [comment]
> Both ValueError and TypeError would typically indicate a bug, though, wouldn't they?

Yes

> It seems like catching them would be much worse than catching SystemExit because it would not distinguish between bugs in RunExecutor and invalid arguments, and thus hide bugs?

ValueError and TypeError are typically raised if you supply invalid arguments to some function. 
E.g. `int("foo")` gives you `ValueError: invalid literal for int() with base 10: 'foo'`, 
`len(1)` gives you `TypeError: object of type 'int' has no len()`. I would say it is the opposite of hiding bugs because it's the way such conditions typically get reported in python. 

> Which cases do you actually want to catch? What is your use case for this?

I do not have a concrete use case for this. I think it would mostly boil down to RunExecutor failed, inform the user, clean up, probably go on with other things. The reason for opening this issue simply was that raising `SystemExit` is very specific for "please exit the interpreter, now" instead of "something went wrong".

> I think that if the caller does everything correctly, RunExecutor should never raise an exception.

As long as you include setting up the environment correctly this is obviously true.

**Comment by PhilippWendler** [comment]
But `runexecutor.main` needs to distinguish between bugs in RunExecutor and incorrect arguments given by the user. The latter should have a nice error message without a stack trace, whereas the former should be visible as bugs and include as much information. This could not be done with `ValueError` etc., because if they are caught and treated as user problems, this would hide bugs.

**Comment by catdog2** [comment]
I would simply subclass them in this case, then you can distinguish them in `runexecutor.main`. Maybe simply, to not overcomplicate this, define your own exception for "illegal arguments" which inherits directly from `Exception` as no one would usually specifically catch a e.g. `ValueError`.

One addition to above: `EnvironmentError` has been [merged](https://www.python.org/dev/peps/pep-3151/#confusing-set-of-os-related-exceptions) into `OSError` (which was a subclass of the former) since python 3.3 so one might want to use or inherit from OSError.
